Add HTLC and atomic swap signing

sisou opened the pull request:
This PR adds two new capabilities to the Keyguard:
- Signing HTLC funding and redeeming transactions for atomic swaps
- Signing HTLC refund transactions for when a swap fails

Signing a swap happens in two steps:
1. The display-relevant data and the involved key paths are handed to the top-level `SignSwap` request. Here the user reviews the swap data and enters their password. Then the Keyguard derives the required private keys at the respective keyPaths for NIM and BTC and stores them in SessionStorage (similar to how we do it during login for the key to derive addresses for discovery in the Hub).
2. After the user confirmed the swap by entering their password, the Hub confirms the swap with Fastspot and lets the Keyguard sign the actual transactions via the `swap-iframe` and it's `SignSwapTransactionsRequest`.

This two step process is necessary, because only when **confirming** a swap with Fastspot does Fastspot provide the necessary data for the HTLC transactions, such as the BTC target address, the BTC HTLC script, Fastspot's refund address (required to create a NIM HTLC) and the common hash (also called `hashRoot` in CoreJS). **However**, once a swap is _confirmed_ with Fastspot, its value counts against the user's 30-day limit. Plus, once confirmed, Fastspot directly locks up their part of the swap, reducing liquidity and incurring transactions fees, which we want to avoid if users have no intention of completing a swap. So we use the entering of the password as the final "yes, let's do this for real" signal of the user. But because we don't want to communicate with Fastspot from inside the Keyguard (no outside connections), we need to delegate the Fastspot interaction back to the Hub. Then, we don't want to require the user to enter their password again, so we store the private keys in SessionStorage to use by a Keyguard iframe. That's why it's a two-step process.

---
~~There is one detail which currently doesn't sit quite right with me about how I implemented it:
I believe there is currently no verification if what is signed in the iframe (in the second step) is actually what the user got displayed and approved in the first step. Meaning, the transactions that get signed with the user's keys _could_ be different from what was shown in the first step.~~

The only way around this, as I see it, is to give _all transaction_ info (except the unavailable ones that only become available with swap confirmation) to the Keyguard in the first step and store this data together with the private keys in SessionStorage. Then in the second step, only accept the missing information necessary to build valid transactions.

**Update:** this is is now how it's implemented.

~~Currently, our Keyguard can only be called from our Hub, where we don't do such weird things, so I think this is fine for now. Albeit obviously not ideal.~~

What do you think?

sisou committed:
WIP Define request typings

sisou committed:
Set up bare-bones sign-swap request

sisou committed:
WIP Require, parse and validate HTLC data/script

sisou committed:
Remove checking of untrusted HTLC data

sisou committed:
Fix signing BTC redeem tx

sisou committed:
Fix signing BTC funding tx

sisou committed:
Fix linter errors

sisou committed:
Always require and validate HTLC data for both swap chains

sisou committed:
Parse swap meta data

And change demo page to use KeyguardClient

sisou committed:
Swap UI layout

sisou committed:
Use sender/recipient label instead of label in nimiqAddresses

sisou committed:
Validate NIM HTLC timeout block height against validityStartHeight

sisou committed:
Allow parsing addresses from strings

sisou committed:
Adjust address display in SignBtcTransaction for longer Bech32 addresses

sisou committed:
Replace swap bar change indicator with SVG for better antialiasing

sisou committed:
Update Swap demo for updated label placement

sisou committed:
Merge branch 'bitcoin' into swap

sisou committed:
Update code for changes from bitcoin branch

sisou committed:
Export BitcoinJS.script

sisou committed:
Add bitcoin-bundle to sign-swap request

sisou committed:
Rework swap to split swap-signing into two requests

sisou committed:
Fix NIM->BTC swaps

sisou committed:
Split service network fees between funding and redeeming ones

They are in different currencies, and the new Fastspot API v1 reports them individually.

sisou committed:
Allow signing BTC HTLC refund transactions via SignBtcTransaction

sisou committed:
Merge branch 'soeren/refund-htlc' into swap

sisou committed:
Client v1.4.0-beta.1 (Bitcoin + Swap methods)

sisou committed:
Improve error for missing stored private keys

sisou committed:
Add check for SessionStorage access before signing a swap

sisou committed:
Minor visual updates for SignSwap UI

sisou committed:
Update translations, fix special character in string

sisou committed:
Merge remote-tracking branch 'origin/master' into swap

sisou committed:
Merge branch 'master' into swap

sisou committed:
Update translations

sisou committed:
Merge branch 'master' into swap

sisou committed:
Convert BitcoinEnabledTopLevelApi into a mixin

To be able to use it in the swap-iframe.

sisou commented on client/src/PublicRequest.ts:
```diff
@@ -26,23 +26,34 @@ export type SingleKeyResult = {
-    recipientType?: Nimiq.Account.Type
-    data?: Uint8Array
+    keyPath: string,
+    senderLabel?: string,
```
I moved the `keyPath` and `senderLabel` into the `TransactionInfo` type to simplify types futher down and when altering/extending them for the swap types.

sisou commented on client/src/PublicRequest.ts:
```diff
@@ -26,23 +26,34 @@ export type SingleKeyResult = {
     outputScript: string,
     value: number,
+    witnessScript?: string,
+    type?: BitcoinTransactionInputType,
```
`witnessScript` is the HTLC script necessary for HTLC redeem and refund transactions. `type` is used to tell the Keyguard which of the two it is, as the witness data the Keyguard has to generate is different with otherwise same request data.

danimoh replied:
This can be typed more precisely:
```ts
export type BitcoinTransactionInput = {
    ...
} & ({
    type?: BitcoinTransactionInputType.DEFAULT,
} | {
    type: BitcoinTransactionInputType.HTLC_REDEEM | BitcoinTransactionInputType.HTLC_REFUND,
    witnessScript: string,
})
```

sisou commented on client/src/PublicRequest.ts:
```diff
@@ -178,6 +185,73 @@ export type SignBtcTransactionRequest
+                | 'validityStartHeight'
+                | 'data'
+                | 'flags'
+            >
```
We don't need quite a bit of information for only displaying the swap info visually, that's why it's all omitted here.

sisou commented on client/src/PublicRequest.ts:
```diff
@@ -178,6 +185,73 @@ export type SignBtcTransactionRequest
+    swapId: string,
+    fund:
+        {type: 'NIM'}
+            & Omit<TransactionInfo, 'keyPath' | 'recipient' | 'recipientType' | 'data'>
```
- `keyPath` was already provided and the private key derived in the top-level request above
- `recipient` is generated automatically for HTLC creation transactions
- `recipientType` is clear, must be `HTLC`
- `data` is omitted because it's optional in `TransactionInfo`, but I am re-adding it below without the optionality.

sisou commented on src/lib/Key.js:
```diff
@@ -28,7 +28,7 @@ class Key {
      */
     derivePublicKey(path) {
-        return Nimiq.PublicKey.derive(this._derivePrivateKey(path));
+        return Nimiq.PublicKey.derive(this.derivePrivateKey(path));
```
I needed to make the `derivePrivateKey` method public, as I need to access the raw private key when signing a swap, to store it in SessionStorage for the iframe.

sisou commented on src/lib/RequestParser.js:
```diff
@@ -188,9 +185,15 @@ class RequestParser { // eslint-disable-line no-unused-vars
     parseAddress(address, name) {
+        if (address === 'CONTRACT_CREATION') {
+            return Nimiq.Address.CONTRACT_CREATION;
+        }
```
For a NIM HTLC creation transaction to work, the recipient must be passed as the _exact_ object reference to `Nimiq.Address.CONTRACT_CREATION`. This can obviously not be done through RPC, so a constant is used and replaced here with the correct object reference.

sisou commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,234 @@
+ * @param {TBase} clazz
+ * @returns {typeof Clazz}
+ */
+function BitcoinRequestParserMixin(clazz) { // eslint-disable-line no-unused-vars
```
This file contains the Bitcoin-related request parsing methods that you already reviewed in the `bitcoin` branch PR. Only minor things changed in the actual methods, which I tried to mark with these review comments.

sisou commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,234 @@
+/**
+ * @template {{}} T
+ * @typedef {new (...args: any[]) => T} Constructor;
+ */
```
This method is taken from official TS docs: https://www.typescriptlang.org/docs/handbook/mixins.html

sisou commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,234 @@
+         * @param {string} name
+         * @returns {'default' | 'htlc-redeem' | 'htlc-refund'}
+         */
+        parseInputType(type, name) {
```
This method is new.

sisou commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,234 @@
+                            (this.parseNonNegativeFiniteNumber(input.value, false, `input[${index}].value`)),
+                        ),
+                    },
+                    type: this.parseInputType(input.type, `input[${index}].type`),
```
`type` is new.

sisou commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,234 @@
+
+                if (input.witnessScript) {
+                    parsed.witnessScript = BitcoinJS.Buffer.from(Nimiq.BufferUtils.fromAny(input.witnessScript));
+                }
```
`witnessScript` parsing is new.

danimoh replied:
Can length / content of outputScript and witnessScript be validated?

sisou replied:
I believe not. They can be totally custom.

sisou commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,234 @@
+         * @param {string} name - name of the property, used in error case only
+         * @returns {string[]}
+         */
+        parseBitcoinPathsArray(paths, name) {
```
This method is new.

sisou commented on src/lib/bitcoin/BitcoinUtils.js:
```diff
@@ -114,4 +116,34 @@ class BitcoinUtils { // eslint-disable-line no-unused-vars
+     * @param {'BIP49' | 'BIP84'} bip
+     * @returns {string}
+     */
+    static addressBytesToAddress(bytes, bip) {
```
This is a left-over from when I did BTC HTLC validation in the Keyguard before the change to the two-step process.

Will probably be re-used once HTLC data is validated again.

sisou commented on src/lib/swap/HtlcUtils.js:
```diff
@@ -0,0 +1,168 @@
+     * @param {unknown} data - Uint8Array
+     * @returns {NimHtlcContents}
+     */
+    static decodeNimHtlcData(data) {
```
The web-offline build of CoreJS does not yet include the HTLC class (https://github.com/nimiq/core-js/pull/572), so we need to recreate its `dataToPlain()` method ourselves to be able to parse NIM HTLC data.

sisou commented on src/lib/swap/HtlcUtils.js:
```diff
@@ -0,0 +1,168 @@
+     * @param {unknown} script - Uint8Array
+     * @returns {BtcHtlcContents}
+     */
+    static decodeBtcHtlcScript(script) {
```
This was fun to figure out :smile:

sisou commented on src/lib/swap/HtlcUtils.js:
```diff
@@ -0,0 +1,168 @@
+        if (!chunks) throw error;
+        const asm = BitcoinJS.script.toASM(chunks).split(' ');
+
+        let branchesVerifiedIndividually = false;
```
A BTC HTLC script can be one of two versions. One where both the redeem and refund branch is verified independently, and one where only one verification happens at the end. This method can parse both. (Should probably have tests...)

nibhar replied:
Agreed. Tests would be good here.

sisou replied:
Test are added in the `oasis` branch: b04b3648f9aaa6f34619a2fd5f5ccfe770db724b

sisou commented on src/lib/swap/HtlcUtils.js:
```diff
@@ -0,0 +1,168 @@
+     * @param {Buffer[]} witness
+     * @returns {Buffer}
+     */
+    static witnessStackToScriptWitness(witness) {
```
This is a re-implementation of a method included in BitcoinJS, which is sadly not exported anywhere.

https://github.com/bitcoinjs/bitcoinjs-lib/blob/533d6c2e6d0aa4111f7948b1c12003cf6ef83137/ts_src/psbt.ts#L1512

sisou commented on src/request/derive-btc-xpub/index.html:
```diff
@@ -71,7 +71,8 @@
 
-                <h1 class="nq-h1" data-i18n="derive-btc-xpub-heading">Add Bitcoin\nto your account</h1>
+                <h1 class="nq-h1" data-i18n="derive-btc-xpub-heading">Add Bitcoin
+                    to your account</h1>
```
This fixes a `\n` character appearing in the translated text on-screen.

sisou commented on src/request/sign-btc-transaction/SignBtcTransaction.js:
```diff
@@ -216,43 +216,82 @@ class SignBtcTransaction {
+            if (parsedInput.type === 'default') {
+                psbt.finalizeInput(i);
+            }
+            if (parsedInput.type === 'htlc-redeem' || parsedInput.type === 'htlc-refund') {
```
The type `htlc-redeem` is currently not used with the standard `SignBtcTransaction` request, but I though it can't hurt to include it, as the code is pretty much the same as for `htlc-refund`, except 2 lines in the middle.

sisou commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -0,0 +1,132 @@
+                    ...request.fund,
+                    flags: Nimiq.Transaction.Flag.CONTRACT_CREATION,
+                    recipient: 'CONTRACT_CREATION',
+                    recipientType: Nimiq.Account.Type.HTLC,
```
Here we set the fixed properties that we _know_ must be present when an HTLC is supposed to be created.

sisou commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,283 @@
+        if (!privateKeys.btc) throw new Error('No BTC key list stored in SessionStorage');
+        if (!privateKeys.btc.length) throw new Error('No BTC keys stored in SessionStorage');
+        if (privateKeys.btc.some(key => !key)) throw new Error('Empty BTC key stored in SessionStorage');
+        if (privateKeys.btc.some(key => key.length !== 64)) throw new Error('Invalid BTC key stored in SessionStorage');
```
These checks are supposed to protect against corrupted or non-existent SessionStorage, and provide the correct error message to help us and the user understand what happened.

sisou commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,283 @@
+        if (privateKeys.btc.some(key => !key)) throw new Error('Empty BTC key stored in SessionStorage');
+        if (privateKeys.btc.some(key => key.length !== 64)) throw new Error('Invalid BTC key stored in SessionStorage');
+
+        // Decode HTLC contents
```
All the commented code is technically TODO, but not required as long as we trust Fastspot and the Hub.

nibhar replied:
As it is already there (just commented), why did you leave it out? It appears important to me to have this working.

sisou commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,283 @@
+                ...request.fund,
+                flags: Nimiq.Transaction.Flag.CONTRACT_CREATION,
+                recipient: 'CONTRACT_CREATION',
+                recipientType: Nimiq.Account.Type.HTLC,
```
Again, we force the three properties that we _know_ must be present for HTLC creation.

sisou committed:
Update wording to match Wallet UI

sisou committed:
Merge branch 'master' into swap

sisou committed:
Update translations

sisou committed:
Client v1.4.0-beta.2

[skip ci]

sisou committed:
Fix displayed BTC swap value swapping to BTC

danimoh commented on client/src/KeyguardClient.ts:
```diff
@@ -33,6 +34,9 @@ import {
     ResultByCommand,
     SignBtcTransactionRequest,
     DeriveBtcXPubRequest,
+    SignSwapRequest,
```
~~`SetupSwapRequest` or `SignSwapSetupRequest` or `ConfirmSwapRequest` could be better names.
Similarly for `SIGN_SWAP`, etc.
But would require many changes...~~

Nah, forget it. It's too many places to change.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -26,23 +26,34 @@ export type SingleKeyResult = {
+    senderType: Nimiq.Account.Type,
+    recipient: Uint8Array,
+    recipientType?: Nimiq.Account.Type,
+    recipientLabel?: string,
```
`recipientLabel` is ignored for `layout: 'checkout'` and `layout: 'cashlink'` and should therefore remain only in the definition of `layout: 'standard'`

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -26,23 +26,34 @@ export type SingleKeyResult = {
 };
 
+export enum BitcoinTransactionInputType {
+    DEFAULT = 'default',
```
Could also be `standard` to be more in line how the default for `layout` is called.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -178,6 +185,73 @@ export type SignBtcTransactionRequest
+export type SignSwapRequest = SimpleRequest & {
+    swapId: string,
+    fund:
+        {type: 'NIM'}
```
Inconsistent spacing around curly brackets in the newly added types.

danimoh commented on client/src/KeyguardClient.ts:
```diff
@@ -162,6 +170,11 @@ export class KeyguardClient {
 
+    public async signSwapTransactions(request: SignSwapTransactionsRequest): Promise<SignSwapTransactionsResult> {
+        const swapIframeBehavior = new SwapIFrameRequestBehavior();
+        return swapIframeBehavior.request(this._endpoint, KeyguardCommand.SIGN_SWAP_TRANSACTIONS, [request]);
```
I would move this logic into `_iframeRequest`.
Otherwise, the type for `_iframeRequest` would need to be updated to exclude the `SignSwapTransactionsRequest` as it's in fact not able to handle that request type.

danimoh commented on demos/Swap.html:
```diff
@@ -0,0 +1,163 @@
+    <script src="../client/dist/KeyguardClient.umd.js"></script>
+    <script src="../node_modules/@nimiq/core-web/web-offline.js"></script>
+    <script src="../src/lib/bitcoin/vendor/NodeBuffer.js"></script>
+    <script src="../src/lib/bitcoin/vendor/BitcoinJS.js"></script>
```
These dependencies are not found.

danimoh commented on demos/Swap.html:
```diff
@@ -0,0 +1,163 @@
+            bitcoinAccount: {
+                balance: 0.001e8, // Sats
+            },
+        });
```
Seems like this is still using the old request.

sisou replied:
The swap is not easily demo-able anymore. I removed the Swap.html demo page.

danimoh commented on src/common.js:
```diff
@@ -55,18 +55,24 @@ async function loadNimiq() {
  * @template {KeyguardRequest.RedirectRequest} T
  * @param {Newable} RequestApiClass - Class object of the API which is to be exposed via RPC
- * @param {object} [options]
+ * @param {Partial<Options>} [opts]
```
Why not mark `loadNimiq` and `whitelist` as `?` instead?

sisou replied:
Because they do exist on the `Options` type (as it has defaults) and would otherwise also be `?` when used in the rest of the function and would require type annotations to force that they are indeed defined.

danimoh commented on src/lib/bitcoin/BitcoinKey.js:
```diff
@@ -23,27 +23,20 @@ class BitcoinKey {
+        switch (BitcoinUtils.parseBipFromDerivationPath(path)) {
+            case BitcoinConstants.BIP.BIP49: address = BitcoinUtils.keyPairToNestedSegwit(keyPair).address; break;
+            case BitcoinConstants.BIP.BIP84: address = BitcoinUtils.keyPairToNativeSegwit(keyPair).address; break;
+            default: break;
```
`default` case can be omitted.

sisou replied:
eslint expects a default case

sisou committed:
Add 1 hour to BTC HTLC timeout to account for Bitcoin mempool rules

sisou committed:
Update tooltip and UI styling for swap fees

nibhar commented on src/request/sign-btc-transaction/SignBtcTransaction.js:
```diff
@@ -216,43 +216,82 @@ class SignBtcTransaction {
         // Sort outputs by value ASC, then address ASC
         outputs.sort((a, b) => (a.value - b.value) || (a.address < b.address ? -1 : 1));
 
-        try {
```
What made this `try {}` block unnecessary?

sisou replied:
I believe I removed it because I'm now sure that it will always work and there are now bugs :D

But I see that there are still lots of places where an Error could be thrown, and having a try-catch wrapper doesn't hurt, so I guess I'll re-add it.

nibhar commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,423 @@
+            );
+        }
+
+        const exchangeRate = Math.round(btcExchangeValue / nimExchangeValue * 1e8) / 1e8;
```
Personally I would prefer to have this as either Btc -> Nim rate or Nim -> Btc rate depending on which direction I swap.

sisou replied:
This is determined by the designs, which have been approved for a long time.

nibhar commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,418 @@
+
+        // Draw distribution graph
+
+        const nimDistributionData = request.nimiqAddresses.map(addressInfo => {
```
Could potentially have been made a component to make this class a little more lightweight.

sisou replied:
Its now extracted in the `oasis` branch: 2950cd5884eb0c9cad55aa1c9ca574810bbbefb2

nibhar commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,418 @@
+        $swapValues.classList.add(`${fundTx.type.toLowerCase()}-to-${redeemTx.type.toLowerCase()}`);
+
+        // Exchange rate
+        const nimExchangeValue = fundTx.type === 'NIM' // eslint-disable-line no-nested-ternary
```
The whole exchange values and exchange rate topic is in my opinion in need of some explanatory comments.

sisou replied:
I added comments, let me know if they explain it well enough. :slightly_smiling_face:

nibhar commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,418 @@
+            ? fundTx.transaction.fee
+            : redeemTx.type === 'NIM'
+                ? redeemTx.transaction.fee
+                : 0) * request.nimFiatRate;
```
So all display within this tooltip is fiat exclusively. I don't know if I like this all that much, as this is also talking about network fees which are (rate-independent) values in crypto.

nibhar commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,418 @@
+     * @param {number} exchangeAmount - In Luna or Satoshi, depending on which currency is funded
+     * @returns {HTMLDivElement}
+     */
+    _makeFeeTooltip(request, exchangeAmount) {
```
Could also be moved to a component to clear this class up.

sisou replied:
Made into a component in the `oasis` branch: ca5d6ba963c5c271b2d74ccea5c5ac728293f7bc

nibhar commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,283 @@
+        /** @type {{nim: string, btc: string[]}} */
+        const privateKeys = JSON.parse(storedPrivateKeys);
+
+        if (!privateKeys.nim) throw new Error('No NIM key stored in SessionStorage');
```
From what I take from the `SignSwapRequest` these do not have to be set specifically.
Only if the fund or the redeem tx is of type Nim this key will be present.

I realise that will always be the case currently, but in the other request it was handled with flexibility in mind while here it is not.

Same for `privateKeys.btc`.

nibhar commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,283 @@
+                // @ts-ignore Argument of type 'import("...").Buffer' is not assignable to parameter of type 'Buffer'.
+                BitcoinJS.Buffer.from(privateKeys.btc[0], 'hex'),
+            );
+            psbt.signInput(0, keyPair);
```
Is there only ever going to be 1 input?

sisou replied:
Yes, BTC HTLCs are only funded with one UTXO.

Doesn't _have_ to, it's just a BTC address, can have multiple UTXOs as well. But the way Fastspot works, there's only one.

nibhar reviewed (changes requested):
Good Job.

From my comments I only deem the one about the commented out todos essential.

sisou committed:
Re-add try...catch wrapper in SignBtcTransaction

sisou committed:
Rework swap request to only sign approved data and validate HTLC scripts

[skip ci]

sisou committed:
Use full Nimiq web.js for swaps

Swaps require the HTLC account type to be included, which it is currently not in the web-offline build. PR pending to add it is nimiq/core-js#572.

sisou committed:
Fix comments

[skip ci]

sisou committed:
Update translations

sisou committed:
Fix dependency check for full core-web/web.js script

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -187,32 +187,52 @@ export type SignBtcTransactionRequest
-        | {type: 'BTC', keyPaths: string[], value: number, fee: number },
-    redeem:
+        & Omit<TransactionInfo,
+            | 'recipient' // Only known in second step (in swap-iframe), derived from htlcData
```
There's nothing before the first `|`.
Is that syntax actually working?

Same also at other places here in this file.

sisou replied:
It's working, I was also surprised. I'm doing it for alignment.

danimoh commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,228 @@
+         * @returns {ParsedBitcoinTransactionInput[]}
+         */
+        parseInputs(inputs) {
+            if (!inputs || !Array.isArray(inputs)) {
```
`!inputs` check already covered by `!Array.isArray()`.

danimoh commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,228 @@
+
+                /** @type {ParsedBitcoinTransactionInput} */
+                const parsed = {
+                    hash: Nimiq.BufferUtils.toHex(Nimiq.BufferUtils.fromAny(input.transactionHash)),
```
Might want to validate the hash length

sisou replied:
An invalid tx hash will fail to get converted into a BTC transaction input and throw eventually, likely after password-entry, but before the request would successfully return.

danimoh commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,228 @@
+         * @returns {string[]}
+         */
+        parseBitcoinPathsArray(paths, name) {
+            if (!paths || !Array.isArray(paths)) {
```
`!paths` part can again be omitted.

danimoh commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,228 @@
+         * @returns {boolean}
+         */
+        isValidBitcoinPath(path) {
+            if (path.match(/^m(\/[0-9]+'?)*$/) === null) return false;
```
Could be more restrictive: require at least purpose', coin_type', account' levels of Bip44 and make sure they are hardened.

Regex can optionally use `\d` instead of `[0-9]` and instead of `path.match(regex)`, `regex.test(path)` could be used.

sisou replied:
This check is taken from Nimiq.ExtendedPrivateKey. The method `parseBitcoinPath` (the only user of this method here) actually checks for BIP49 vs BIP84.

sisou replied:
I think the Keyguard should make no assumption about the derivation path used, to also allow non-standard usage. Does the BTC Ledger app restrict the paths?

danimoh commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,228 @@
+            // Overflow check.
+            const segments = path.split('/');
+            for (let i = 1; i < segments.length; i++) {
+                if (!Nimiq.NumberUtils.isUint32(parseInt(segments[i], 10))) return false;
```
Indices must additionally be `< 2 ** 31` according to Bip32.

Loop can optionally also be simplified using `for (const ... of ...)` or `segments.some()`.

danimoh commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,228 @@
+                    /** @type {{address: unknown}} */ (output).address,
+                    `${parameterName}.address`,
+                ),
+                label: this.parseLabel(/** @type {{label: unknown}} */ (output).label),
```
Parameter name not specified here.

danimoh commented on src/lib/bitcoin/BitcoinRequestParserMixin.js:
```diff
@@ -0,0 +1,228 @@
+                return undefined;
+            }
+
+            if (!output || typeof output !== 'object') {
```
`!output` check is redundant.
Same in `parseChangeOutput`.

sisou replied:
Actually, `!output` checks for `null`, which is `typeof 'object'`.

danimoh commented on src/lib/bitcoin/BitcoinUtils.js:
```diff
@@ -114,4 +116,34 @@ class BitcoinUtils { // eslint-disable-line no-unused-vars
+
+        if (bip === BitcoinConstants.BIP.BIP49) {
+            return BitcoinJS.address.toBase58Check(
+                // @ts-ignore Argument of type 'Uint8Array | Buffer' is not assignable to parameter of type 'Buffer'.
```
Could be avoided by actually ensuring that bytes is a Buffer via `bytes = Buffer.from()`

sisou replied:
That only changes the type error to `Type Buffer of import(...) is not assignable to type Buffer`, because the type `Buffer` that is required is the NodeJS `Buffer`, but the library for the browser is a different `Buffer`.

danimoh commented on src/lib/swap/HtlcUtils.js:
```diff
@@ -0,0 +1,173 @@
+        const hashCount = buf.readUint8();
+        const timeout = buf.readUint32();
+
+        if (!sender || !recipient || !hashRoot || !timeout) {
```
I think this check is never true.

sisou replied:
You are right, the length of the buffer is checked before, so all deserialization will succeed.

danimoh commented on src/lib/swap/HtlcUtils.js:
```diff
@@ -0,0 +1,173 @@
+
+        let branchesVerifiedIndividually = false;
+
+        /* eslint-disable no-plusplus */
```
Feel free to also globally allow usage of `++`.

danimoh commented on src/lib/swap/HtlcUtils.js:
```diff
@@ -0,0 +1,173 @@
+
+        // Check hash
+        if (asm[++i] !== 'OP_SHA256' || asm[i + 2] !== 'OP_EQUALVERIFY') throw error;
+        const hash = Nimiq.BufferUtils.toHex(Nimiq.BufferUtils.fromHex(asm[++i]));
```
Checking hash length might be a good idea.

Same for all other data read from the script.

And I assume the `fromHex` followed by `toHex` is for verification purposes?

sisou replied:
I added a hash length check. And yes, the fromHex(toHex()) is for verification that it is valid hex data.

danimoh commented on src/lib/swap/HtlcUtils.js:
```diff
@@ -0,0 +1,173 @@
+        // Bitcoin HTLC timeouts are backdated 1 hour, to account for Bitcoin's
+        // minimum age for valid transaction locktimes (6 blocks).
+        // @ts-ignore Argument of type 'Buffer' is not assignable to parameter of type 'Buffer'
+        const timeoutTimestamp = BitcoinJS.script.number.decode(BitcoinJS.Buffer.from(asm[++i], 'hex')) + (60 * 60);
```
Does 60 * 60 stand for 60 minutes with 60 seconds each, so is the timestamp in seconds?

sisou replied:
Yes and yes.

danimoh commented on src/lib/swap/HtlcUtils.js:
```diff
@@ -0,0 +1,173 @@
+         * @param {Buffer} slice
+         */
+        function writeSlice(slice) {
+            buffer = buffer.concat([...slice.subarray(0)]);
```
What's the `subarray` for?
Could also use spread syntax: `buffer = [...buffer, ...slice];`

sisou replied:
It's to convert the Buffer into an Uint8Array.

danimoh commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -0,0 +1,142 @@
+                type: 'BTC',
+                inputs: this.parseInputs(request.fund.inputs),
+                recipientOutput: {
+                    value: this.parsePositiveInteger(request.fund.recipientOutput.value),
```
`parameterName` not set.
Might also apply to some other things parsed here.

sisou replied:
I believe all other checks have the parameter name set.

danimoh commented on src/request/sign-swap/SignSwapApi.js:
```diff
@@ -0,0 +1,142 @@
+            throw new Errors.InvalidRequestError(
+                'The address details of the NIM address doing the swap must be provided',
+            );
+        }
```
Could also validate that the balance is `>=` the funding amount.

nibhar commented on src/request/sign-btc-transaction/SignBtcTransaction.js:
```diff
@@ -158,65 +158,70 @@ class SignBtcTransaction {
+                const keyPair = btcKey.deriveKeyPair(input.keyPath);
+                const output = BitcoinUtils.keyPairToNativeSegwit(keyPair).output;
+                if (!output) {
+                    TopLevelApi.setLoading(false);
```
Unnecessary as the catch block also does it.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,492 @@
+        // Exchange rate
+        const nimExchangeValue = fundTx.type === 'NIM' // eslint-disable-line no-nested-ternary
+            // When the user funds NIM, the service receives the HTLC balance - their network fee.
+            ? (fundTx.transaction.value - request.serviceFundingNetworkFee) / 1e5
```
Could use `lunasToCoins`.
Same below.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,492 @@
+
+        const btcExchangeValue = fundTx.type === 'BTC' // eslint-disable-line no-nested-ternary
+            // When the user funds BTC, the service receives the HTLC balance - their network fee.
+            ? (fundTx.recipientOutput.value - request.serviceFundingNetworkFee) / 1e8
```
Could use `satoshisToCoins`.
Same below.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,492 @@
+            : BitcoinUtils.satoshisToCoins(serviceExchangeFee) * request.btcFiatRate;
+
+        const theirExchangeFeePercentage = NumberFormatting.formatNumber(
+            request.serviceExchangeFee / (exchangeAmount - request.serviceExchangeFee) * 100,
```
Can use `serviceExchangeFee` instead of `request.serviceExchangeFee`

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,492 @@
+            const keyPairs = request.fund.inputs.map(input => btcKey.deriveKeyPair(input.keyPath));
+            const privKeys = keyPairs.map(keyPair => Nimiq.BufferUtils.toHex(
+                /** @type {Buffer} */ (keyPair.privateKey),
+            ));
```
Could also directly do `toString('hex')` on the `Buffer`.
Same below.

nibhar commented on src/request/sign-btc-transaction/SignBtcTransaction.js:
```diff
@@ -226,17 +231,62 @@ class SignBtcTransaction {
-            psbt.finalizeAllInputs();
+            let i = 0;
+            for (const parsedInput of request.inputs) {
+                if (parsedInput.type === 'default') {
```
I assume there are more types than the 3 that we are currently using. If we are likely going to be using more in the future i would prefer this being a switch:
```js
switch (parsedInput.type) {
  case 'default':
    psbt.finalizeInput(i);
    break;
  case 'htlc-redeem':
  case 'htlc-refund':
    ...
    break;
  default:
    throw new Errors.KeyguardError('Unsupported input type');
}
```
I realise that the Parser takes care of different types and the default arm is technically unreachable.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,492 @@
+                            } : {}
+                        ),
+                    },
+                }),
```
This can be simplified passing a custom replacer to `JSON.stringify` which serializes transactions and buffers.

danimoh commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,492 @@
+    }
+}
+
+SignSwap.Pages = {
+    CONFIRM_SWAP: 'confirm-swap',
+};
```
As there is only a single page, is this necessary?

sisou replied:
I think it's convention, and is also used for the DOM selector.

danimoh commented on src/request/sign-swap/index.html:
```diff
@@ -0,0 +1,156 @@
+          child-src 'self' blob:;
+          style-src 'self';
+          font-src 'self';
+          connect-src 'self' https://api.coingecko.com">
```
I think coingecko is currently not used in the sign-swap request.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+        const storedData = sessionStorage.getItem(storageKey);
+        sessionStorage.removeItem(storageKey); // Delete storage
+
+        if (!storedData) throw new Error('No swap stored in SessionStorage');
```
In error case should delete all stored swap data to avoid the swap id being brute-forceable.
Note that the session storage can be iterated with https://developer.mozilla.org/en-US/docs/Web/API/Storage.key in combination with `sessionStorae.length` to find the entries to delete. But apparently also `Object.keys(sessionStorage)` seems to work.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+        if (request.fund.type === 'BTC' || request.redeem.type === 'BTC') {
+            if (!privateKeys.btc) throw new Error('No BTC key list stored in SessionStorage');
+            if (!privateKeys.btc.length) throw new Error('No BTC keys stored in SessionStorage');
+            if (privateKeys.btc.some(key => !key)) throw new Error('Empty BTC key stored in SessionStorage');
+            if (privateKeys.btc.some(key => key.length !== 64)) {
+                throw new Error('Invalid BTC key stored in SessionStorage');
+            }
```
These could be combined.

sisou replied:
I had them separate for the different type of error message, which I though useful for debugging.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+
+        // Parse request
+        /** @type {Parsed<KeyguardRequest.SignSwapTransactionsRequest>} */
+        // @ts-ignore Missing the following properties: fund, redeem
```
To avoid this, you can probably also first set `fund` and `redeem` as variables and then later do
```js
const parsedRequest = {
    fund,
    redeem,
    swapId: request.swapId,
};
```

nibhar replied:
I am not too fond of this `@ts-ignore`.
Maybe assemble the parts first and the `parsedRequest` last?

nibhar commented on src/request/sign-swap/SignSwap.js:
```diff
@@ -0,0 +1,492 @@
+
+                    // Serialize request to store in SessionStorage
+                    request: {
+                        ...request,
```
In my opinion a little overboard with inlining.

i.e
```js
let copiedRequest = Object.assign({}, request);
if (copiedRequest.fund.type === 'NIM') {
    copiedRequest.fund.transaction = request.fund.transaction.toPlain();
}
```
Or something along those lines would have worked too. Potentially even in place. In my opinion is easier to read and shorter, as well.

sisou replied:
I was actually quite proud of this construct :laughing:

nibhar commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+    }
+}
+
+SwapIFrameApi.SESSION_STORAGE_KEY_PREFIX = 'swap_id_';
```
This should be put into `Constants.js` as it is duplicated for `SignSwapApi.js` as the "same" constant.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+            }
+
+            const htlcAddress = BitcoinJS.payments.p2wsh({
+                // @ts-ignore Type 'Uint8Array' is not assignable to type 'Buffer'.
```
Are you still getting this warning even though you're doing `Buffer.from`?

sisou replied:
No, actually the error about Buffer and the polyfilled Buffer being not the same.

nibhar commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+
+        // Deserialize stored request
+        /** @type {Parsed<KeyguardRequest.SignSwapRequest>} */
+        const storedRequest = {
```
Again this could be done with less inlining for a more readable code in my opinion.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+                type: 'NIM',
+                htlcDetails,
+                htlcData: request.redeem.htlcData,
+                htlcAddress: this.parseAddress(request.redeem.htlcAddress, 'redeem.htlcAddress')
+                    .toUserFriendlyAddress(),
```
Could also be calculated from the actual contract creation tx, which would thus need to be passed as input.

sisou replied:
I think that is unnecessary because if the HTLC address actually isn't this address, the swap does not get past step 2 in the wallet.

And as a precaution against the HTLC address getting faked in the request: a whole transaction can just as easily be faked - the Keyguard cannot validate if a transaction is included in the chain or not.

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+        }
+
+        if (request.fund.type === 'BTC' && storedRequest.fund.type === 'BTC') {
+            const htlcDetails = HtlcUtils.decodeBtcHtlcScript(request.fund.htlcScript);
```
The timeout timestamp could be roughly compared to our local time.

nibhar commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+            };
+        }
+
+        if (parsedRequest.redeem.type === 'NIM' && storedRequest.redeem.type === 'NIM') {
```
Pretty much the same as the funding except for about 2 lines. Can maybe be simplified into a single block.

nibhar reviewed (commented):
Only comments about some minor things, which also in part come down to preference.

Nice work!

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+            const psbt = new BitcoinJS.Psbt({ network: BitcoinUtils.Network });
+
+            // Add inputs
+            // @ts-ignore Argument of type 'Uint8Array' is not assignable to parameter of type 'Buffer'.
```
Does `Buffer.from` help?

sisou replied:
No, because the browser-polyfilled "Buffer" (which is a Uint8Array internally) is still not the same as the _type_ `Buffer`.

sisou replied:
And making the type use `Buffer` instead, just shifts the TS errors from this here usage to the creation of the object, plus it would be using non-browser types.

sisou replied:
Maybe this would be cleaner though. I opened https://github.com/nimiq/keyguard/issues/432 to investigate this after this is merged.

sisou committed:
Merge branch 'master' into swap

sisou committed:
Update translations

sisou committed:
Client v1.4.0

danimoh commented on src/request/swap-iframe/SwapIFrameApi.js:
```diff
@@ -0,0 +1,385 @@
+
+            // Add change output
+            if (storedRequest.fund.changeOutput) {
+                // The address is set in SignSwap after password entry.
```
Could also validate this, doesn't hurt.
Same below.

danimoh commented on src/request/swap-iframe/index.js:
```diff
@@ -0,0 +1,9 @@
+/* global SwapIFrameApi */
+
+runKeyguard(SwapIFrameApi, {
+    loadNimiq: false,
```
I think it's always required?
So could also be true.

sisou replied:
Not if we use this iframe for other swaps in the future as well, e.g EUR-BTC.

danimoh commented on src/translations/nl.json:
```diff
@@ -0,0 +1,219 @@
+{
```
As Dutch is not being released as part of this PR, it should also not be added here.

danimoh commented on tools/build.sh:
```diff
@@ -120,7 +121,7 @@ fi
 
 # For bitcoinjs-lib v5.2.0 and Buffer v5.6.0
-bitcoinjs_hashsum="9cb30b789ff1a79cc1925d7f49ee827b2e9eb89531a58c51ca6ca996b15597eb  src/lib/bitcoin/BitcoinJS.js"
+bitcoinjs_hashsum="5a02ae59046a7ee4f386f5828097aa21bdf52657acca4ae472d5e89e8332ba43  src/lib/bitcoin/BitcoinJS.js"
```
What changed about BitcoinJS?

sisou replied:
Had to add the `ECPair` to the exports.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -178,6 +185,93 @@ export type SignBtcTransactionRequest
+    ) | (
+        {type: 'BTC'}
+        & Omit<BitcoinTransactionInfo,
+            | 'recipientOutput' // Replaced below
```
Could also use `Transform` helper type from `Keyguard.d.ts`

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -187,32 +187,52 @@ export type SignBtcTransactionRequest
-            & { senderLabel: string }
-        | {type: 'BTC', keyPaths: string[], value: number, fee: number },
-    redeem:
+        & Omit<TransactionInfo,
```
Using `Pick` instead of `Omit` would probably be shorter and give a better idea of what's actually included.

sisou replied:
The type has 11 properties, so `Pick` would be longer.

It would become this:
```
        & Pick<TransactionInfo,
            | 'keyPath'
            | 'senderLabel'
            | 'sender'
            | 'senderType'
            | 'value'
            | 'fee'
            | 'validityStartHeight'
        >
```

But it would loose all the explanation of why the other ones are omitted. I prefer the `Omit<>` method.

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -187,32 +187,52 @@ export type SignBtcTransactionRequest
+        & { senderLabel: string }
+    ) | (
+        {type: 'BTC'}
+        & Omit<BitcoinTransactionInfo,
```
The fee is not validated in the Keyguard which is not ideal.

sisou replied:
How should we validate the fee?

danimoh commented on client/src/PublicRequest.ts:
```diff
@@ -187,32 +187,52 @@ export type SignBtcTransactionRequest
-            >
-            & { recipientLabel: string }
-        | {type: 'BTC', keyPaths: string[], value: number, fee: number },
+        & Omit<TransactionInfo,
```
Could use `Pick` again.

sisou replied:
See https://github.com/nimiq/keyguard/pull/418#discussion_r672147730

danimoh committed:
Merge branch 'master' into swap

sisou committed:
Handle @nibhar's PR review

sisou committed:
First batch of @danimoh's PR review

sisou committed:
Merge branch 'master' into swap

sisou committed:
Update translations

sisou committed:
Handle PR review comments
